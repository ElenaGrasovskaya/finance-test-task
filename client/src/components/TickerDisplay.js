import React from 'react';
import { BsArrowDownSquareFill } from "react-icons/bs";
import { BsArrowUpSquareFill } from "react-icons/bs";
import { TickerContainer, StyledIcon, StyledBlock, TickerName, StyledBlockLast } from "../styledComponents/TickerDisplayStyles"

function TickerDisplay(ticker) {
  const display = ticker.ticker;
  return (
    <TickerContainer>
      <StyledIcon mainColor={(Number(display.change) > 150)}>{(Number(display.change) > 150) ? <BsArrowUpSquareFill /> : <BsArrowDownSquareFill />}</StyledIcon>
      <StyledBlock><TickerName>{display.ticker}</TickerName><div>{display.price}</div></StyledBlock>
      <StyledBlockLast><p>{display.change}</p><p>{display.change_percent}%</p></StyledBlockLast>
    </TickerContainer>
  )
}

export default TickerDisplay

