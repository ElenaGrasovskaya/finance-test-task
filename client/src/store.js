import { applyMiddleware, combineReducers } from 'redux'
import thunkMiddleware from 'redux-thunk'
import { composeWithDevTools } from 'redux-devtools-extension'
import { configureStore } from '@reduxjs/toolkit'
import { tickerListReducer } from './reducers/tickerReducers';

const reducer = combineReducers({
    tickerList: tickerListReducer,
})

const middlewares = [thunkMiddleware]
const middlewareEnhancer = applyMiddleware(...middlewares)
const enhancers = [middlewareEnhancer]
const composedEnhancers = composeWithDevTools(...enhancers)

const store = configureStore({ reducer }, composedEnhancers)

export default store