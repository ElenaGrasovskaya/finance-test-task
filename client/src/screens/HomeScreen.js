import React, { useEffect, useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import {
    StyledContainer,
    StyledHeader,
    TickerContainer,
    CloseButton, RefreshButton,
    IntervalForm,
    TickerDisplayButton
} from "../styledComponents/HomeScreenStyles"
import { listTickers } from '../actions/tickerActions'
import { IoMdSync } from "react-icons/io";
import TickerDisplay from '../components/TickerDisplay';
import { AiFillCloseCircle } from "react-icons/ai";
import { displayTickersTemplate } from '../constants/tickerConstants';
import TickerSingleChart from "../components/TickerSingleChart";


function HomeScreen() {

    const [hideList, setHideList] = useState([]);
    const [selectedTicker, setSelectedTicker] = useState("AAPL");
    const [refreshInterval, setRefreshInterval] = useState(5000)
    const [updateInterval, setUpdateInterval] = useState(5000);
    const [displayTickers, setDisplayTickers] = useState([...displayTickersTemplate])

    const dispatch = useDispatch();
    const tickerList = useSelector((state) => state.tickerList);
    const { tickers, loading, error, allTickers } = tickerList;


    
    useEffect(() => {
        dispatch(listTickers(true));
        
        const interval = setInterval(() => {
      
            setDisplayTickers(tickers);
            
        }, updateInterval);

        return () => {
            dispatch(listTickers(false));
            clearInterval(interval);
        }

    }, [updateInterval, displayTickers, dispatch]);


    

    const handleHideTicker = (ticker) => {
        const newHideList = new Set([...hideList, ticker]);
        setHideList([...newHideList])
    }

    const handleSelectTicker = (ticker) => {
        console.log(ticker);
        setSelectedTicker(ticker);
    }

    return (

        <div>
            <StyledHeader>INCODE Finance Controll</StyledHeader>
            {(!loading && !error) ? (<StyledContainer>
                {
                    tickers.map((ticker, index) => (!hideList.find((el) => el === ticker.ticker)) ? (

                        <TickerContainer key={index + 40}><CloseButton key={index + 10} onClick={(e) => handleHideTicker(ticker.ticker)}><AiFillCloseCircle key={index + 30} /></CloseButton>
                            <TickerDisplayButton onClick={(e)=>{handleSelectTicker(ticker.ticker)}}><TickerDisplay ticker={ticker} key={index + 20} /></TickerDisplayButton></TickerContainer>) : ("")
                    )
                }

            </StyledContainer>) : (<StyledContainer>
                {
                    displayTickersTemplate.map((ticker, index) => (!hideList.find((el) => el === ticker.ticker)) ? (

                        <TickerContainer key={index + 40}><CloseButton key={index + 10} onClick={(e) => handleHideTicker(ticker.ticker)}><AiFillCloseCircle key={index + 30} /></CloseButton>
                            <TickerDisplayButton onClick={(e)=>{handleSelectTicker(ticker.ticker)}}><TickerDisplay ticker={ticker} key={index + 20} /></TickerDisplayButton></TickerContainer>) : ("")
                    )
                }

            </StyledContainer>)}

            <IntervalForm onSubmit={(e) => {
                e.preventDefault();

                setUpdateInterval(refreshInterval);

            }}>
                <RefreshButton onClick={(e) => { e.preventDefault(); setHideList([]); }}><IoMdSync /></RefreshButton>
                <label htmlFor='interval'>Set new refresh interval: </label>
                <input
                    type="number"
                    value={refreshInterval}
                    onChange={(e) => {
                        setRefreshInterval(Number(e.target.value));
                    }}
                    name="interval"
                    id="interval"
                ></input>

                <button type="submit">Apply
                </button>
            </IntervalForm>
            <TickerSingleChart tickers={displayTickersTemplate} allTickers={allTickers} selectedTicker = {selectedTicker} />


        </div>

    )
}

export default HomeScreen

/*
 {(!loading && !error) ? <TickerLineChart tickers={tickers} loading={loading} error={error} allTickers={allTickers} /> : (<h2>Loading...</h2>)}


*/


