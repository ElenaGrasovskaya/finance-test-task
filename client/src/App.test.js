import { render, screen } from '@testing-library/react';
import TickerDisplay from './components/TickerDisplay'
import HomeScreen from './screens/HomeScreen'
import { fakeTicker } from './testUtils/testUtils';
import {Provider} from 'react-redux';
import store from './store'
const tickers = fakeTicker();

describe('<TickerDisplay/>', () => {
    it('renders out the ticker title', () => {
        const { container , debug } = render(<TickerDisplay ticker={tickers[0]}/>);
        debug();
        expect(screen.getByText("AAPL")).toBeInTheDocument(); 
                
    });

    it('renders and matches a snapshot', () => {
      const { container , debug } = render(<TickerDisplay ticker={tickers[0]}/>);
      debug();
      expect(container).toMatchSnapshot();

  });
     
});

describe('<HomeScreen/>', () => {
  it('renders out the button for changing refresh interval', () => {
      const { container , debug } = render(<Provider store={store}><HomeScreen/></Provider>);
      debug();
      expect(screen.getByText("Apply")).toBeInTheDocument();
      
  });

  it('renders and matches a snapshot', () => {
    const { container , debug } = render(<Provider store={store}><HomeScreen/></Provider>);
    debug();
    expect(container).toMatchSnapshot();

});
   
});