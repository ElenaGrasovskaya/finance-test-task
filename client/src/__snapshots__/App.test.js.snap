// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomeScreen/> renders and matches a snapshot 1`] = `
<div>
  <div>
    <h1
      class="sc-gKXOVf dZDvWN"
    >
      INCODE Finance Controll
    </h1>
    <h2>
      Loading...
    </h2>
    <form
      class="sc-jqUVSM jcNDAj"
    >
      <button
        class="sc-papXJ cJebiC"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 93.09V32l-80 81.454 80 81.456v-61.093c65.996 0 120 54.982 120 122.183 0 20.363-5 39.714-14.004 57.016L391 342.547c15.996-25.457 25-54.988 25-86.547 0-89.599-72.002-162.91-160-162.91zm0 285.094c-66.001 0-120-54.988-120-122.184 0-20.363 5-39.709 13.999-57.02L121 169.454C104.999 193.89 96 224.436 96 256c0 89.599 72.002 162.91 160 162.91V480l80-81.453-80-81.457v61.094z"
          />
        </svg>
      </button>
      <label
        for="interval"
      >
        Set new refresh interval: 
      </label>
      <input
        id="interval"
        name="interval"
        type="number"
        value="5000"
      />
      <button
        type="submit"
      >
        Apply
      </button>
    </form>
    <div
      class="k-chart k-widget"
      style="height: 2000px; position: relative;"
    >
      <div
        class="k-chart-surface"
        style="position: relative;"
      >
        <svg
          style="width: 100%; height: 100%; overflow: hidden;"
          version="1.1"
          xmlns="http://www.w3.org/2000/svg"
          xmlns:xlink="http://www.w3.org/1999/xlink"
        >
          <defs />
          <g>
            <path
              d="M0 0 L 600 0 600 400 0 400Z"
              fill="none"
              stroke="none"
            />
            <path
              d="M24 5 L 594 5 594 386 24 386Z"
              fill="#fff"
              fill-opacity="0"
              stroke="none"
            />
            <g>
              <g />
              <g>
                <path
                  d="M24.5 310.5 L 594.5 310.5"
                  fill="none"
                  stroke="none"
                />
                <path
                  d="M24.5 234.5 L 594.5 234.5"
                  fill="none"
                  stroke="none"
                />
                <path
                  d="M24.5 157.5 L 594.5 157.5"
                  fill="none"
                  stroke="none"
                />
                <path
                  d="M24.5 81.5 L 594.5 81.5"
                  fill="none"
                  stroke="none"
                />
                <path
                  d="M24.5 5.5 L 594.5 5.5"
                  fill="none"
                  stroke="none"
                />
              </g>
              <g>
                <g>
                  <g>
                    <path
                      d="M24.5 386.5 L 594.5 386.5"
                      fill="none"
                      stroke="none"
                    />
                    <path
                      d="M24.5 386.5 L 24.5 390.5"
                      fill="none"
                      stroke="none"
                    />
                  </g>
                </g>
              </g>
              <g>
                <g>
                  <path
                    d="M24.5 5.5 L 24.5 386.5"
                    fill="none"
                    stroke="black"
                    stroke-width="1"
                  />
                  <path
                    d="M20.5 386.5 L 24.5 386.5"
                    fill="none"
                    stroke="black"
                    stroke-width="1"
                  />
                  <path
                    d="M20.5 310.5 L 24.5 310.5"
                    fill="none"
                    stroke="black"
                    stroke-width="1"
                  />
                  <path
                    d="M20.5 234.5 L 24.5 234.5"
                    fill="none"
                    stroke="black"
                    stroke-width="1"
                  />
                  <path
                    d="M20.5 157.5 L 24.5 157.5"
                    fill="none"
                    stroke="black"
                    stroke-width="1"
                  />
                  <path
                    d="M20.5 81.5 L 24.5 81.5"
                    fill="none"
                    stroke="black"
                    stroke-width="1"
                  />
                  <path
                    d="M20.5 5.5 L 24.5 5.5"
                    fill="none"
                    stroke="black"
                    stroke-width="1"
                  />
                </g>
              </g>
              <g>
                <text
                  fill="black"
                  fill-opacity="1"
                  stroke="none"
                  style="font:  ;white-space:pre;"
                  x="16"
                  y="387"
                >
                  0
                </text>
              </g>
              <g>
                <text
                  fill="black"
                  fill-opacity="1"
                  stroke="none"
                  style="font:  ;white-space:pre;"
                  x="16"
                  y="310.8"
                >
                  100
                </text>
              </g>
              <g>
                <text
                  fill="black"
                  fill-opacity="1"
                  stroke="none"
                  style="font:  ;white-space:pre;"
                  x="16"
                  y="234.6"
                >
                  200
                </text>
              </g>
              <g>
                <text
                  fill="black"
                  fill-opacity="1"
                  stroke="none"
                  style="font:  ;white-space:pre;"
                  x="16"
                  y="158.4"
                >
                  300
                </text>
              </g>
              <g>
                <text
                  fill="black"
                  fill-opacity="1"
                  stroke="none"
                  style="font:  ;white-space:pre;"
                  x="16"
                  y="82.2"
                >
                  400
                </text>
              </g>
              <g>
                <text
                  fill="black"
                  fill-opacity="1"
                  stroke="none"
                  style="font:  ;white-space:pre;"
                  x="16"
                  y="6"
                >
                  500
                </text>
              </g>
              <g
                transform="matrix(0,-1,1,0,-191,206)"
              >
                <text
                  fill="black"
                  fill-opacity="1"
                  stroke="none"
                  style="font:  ;white-space:pre;"
                  x="10"
                  y="197"
                >
                  Ticker price
                </text>
              </g>
            </g>
            <g />
          </g>
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`<TickerDisplay/> renders and matches a snapshot 1`] = `
<div>
  <div
    class="sc-bczRLJ gkYZus"
  >
    <div
      class="sc-eCYdqJ Sqffu"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M2 0a2 2 0 0 0-2 2v12a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V2a2 2 0 0 0-2-2H2zm6.5 4.5v5.793l2.146-2.147a.5.5 0 0 1 .708.708l-3 3a.5.5 0 0 1-.708 0l-3-3a.5.5 0 1 1 .708-.708L7.5 10.293V4.5a.5.5 0 0 1 1 0z"
        />
      </svg>
    </div>
    <div
      class="sc-dkzDqf gzSwhT"
    >
      <div
        class="sc-gsnTZi jRNdvI"
      >
        AAPL
      </div>
      <div>
        279.71
      </div>
    </div>
    <div
      class="sc-hKMtZM jCLcKn"
    >
      <p>
        65.07
      </p>
      <p>
        0.79
        %
      </p>
    </div>
  </div>
</div>
`;

exports[`<TickerLineChart/> renders and matches a snapshot 1`] = `
<div>
  <div
    class="k-chart k-widget"
    style="height: 2000px; position: relative;"
  >
    <div
      class="k-chart-surface"
      style="position: relative;"
    >
      <svg
        style="width: 100%; height: 100%; overflow: hidden;"
        version="1.1"
        xmlns="http://www.w3.org/2000/svg"
        xmlns:xlink="http://www.w3.org/1999/xlink"
      >
        <defs />
        <g>
          <path
            d="M0 0 L 600 0 600 400 0 400Z"
            fill="none"
            stroke="none"
          />
          <path
            d="M24 5 L 594 5 594 386 24 386Z"
            fill="#fff"
            fill-opacity="0"
            stroke="none"
          />
          <g>
            <g />
            <g>
              <path
                d="M24.5 310.5 L 594.5 310.5"
                fill="none"
                stroke="none"
              />
              <path
                d="M24.5 234.5 L 594.5 234.5"
                fill="none"
                stroke="none"
              />
              <path
                d="M24.5 157.5 L 594.5 157.5"
                fill="none"
                stroke="none"
              />
              <path
                d="M24.5 81.5 L 594.5 81.5"
                fill="none"
                stroke="none"
              />
              <path
                d="M24.5 5.5 L 594.5 5.5"
                fill="none"
                stroke="none"
              />
            </g>
            <g>
              <g>
                <g>
                  <path
                    d="M24.5 386.5 L 594.5 386.5"
                    fill="none"
                    stroke="none"
                  />
                  <path
                    d="M24.5 386.5 L 24.5 390.5"
                    fill="none"
                    stroke="none"
                  />
                </g>
              </g>
            </g>
            <g>
              <g>
                <path
                  d="M24.5 5.5 L 24.5 386.5"
                  fill="none"
                  stroke="black"
                  stroke-width="1"
                />
                <path
                  d="M20.5 386.5 L 24.5 386.5"
                  fill="none"
                  stroke="black"
                  stroke-width="1"
                />
                <path
                  d="M20.5 310.5 L 24.5 310.5"
                  fill="none"
                  stroke="black"
                  stroke-width="1"
                />
                <path
                  d="M20.5 234.5 L 24.5 234.5"
                  fill="none"
                  stroke="black"
                  stroke-width="1"
                />
                <path
                  d="M20.5 157.5 L 24.5 157.5"
                  fill="none"
                  stroke="black"
                  stroke-width="1"
                />
                <path
                  d="M20.5 81.5 L 24.5 81.5"
                  fill="none"
                  stroke="black"
                  stroke-width="1"
                />
                <path
                  d="M20.5 5.5 L 24.5 5.5"
                  fill="none"
                  stroke="black"
                  stroke-width="1"
                />
              </g>
            </g>
            <g>
              <text
                fill="black"
                fill-opacity="1"
                stroke="none"
                style="font:  ;white-space:pre;"
                x="16"
                y="387"
              >
                0
              </text>
            </g>
            <g>
              <text
                fill="black"
                fill-opacity="1"
                stroke="none"
                style="font:  ;white-space:pre;"
                x="16"
                y="310.8"
              >
                100
              </text>
            </g>
            <g>
              <text
                fill="black"
                fill-opacity="1"
                stroke="none"
                style="font:  ;white-space:pre;"
                x="16"
                y="234.6"
              >
                200
              </text>
            </g>
            <g>
              <text
                fill="black"
                fill-opacity="1"
                stroke="none"
                style="font:  ;white-space:pre;"
                x="16"
                y="158.4"
              >
                300
              </text>
            </g>
            <g>
              <text
                fill="black"
                fill-opacity="1"
                stroke="none"
                style="font:  ;white-space:pre;"
                x="16"
                y="82.2"
              >
                400
              </text>
            </g>
            <g>
              <text
                fill="black"
                fill-opacity="1"
                stroke="none"
                style="font:  ;white-space:pre;"
                x="16"
                y="6"
              >
                500
              </text>
            </g>
            <g
              transform="matrix(0,-1,1,0,-191,206)"
            >
              <text
                fill="black"
                fill-opacity="1"
                stroke="none"
                style="font:  ;white-space:pre;"
                x="10"
                y="197"
              >
                Ticker price
              </text>
            </g>
          </g>
          <g />
        </g>
      </svg>
    </div>
  </div>
</div>
`;
